support/controlplane-component/controlplane-component.go

```diff
@@ -311,57 +278,14 @@ func (c *controlPlaneWorkload) reconcileWorkload(cpContext ControlPlaneContext)
 		}
 	}
 
-	switch c.workloadType {
-	case deploymentWorkloadType:
-		if err := c.applyOptionsToDeployment(cpContext, workloadObj.(*appsv1.Deployment), oldWorkloadObj.(*appsv1.Deployment)); err != nil {
-			return err
-		}
-	case statefulSetWorkloadType:
-		if err := c.applyOptionsToStatefulSet(cpContext, workloadObj.(*appsv1.StatefulSet), oldWorkloadObj.(*appsv1.StatefulSet)); err != nil {
-			return err
-		}
-	}
-
-	if _, err := cpContext.CreateOrUpdateV2(cpContext, cpContext.Client, workloadObj); err != nil {
-		return err
-	}
-	return nil
-}
-
-func (c *controlPlaneWorkload) applyOptionsToDeployment(cpContext ControlPlaneContext, deployment *appsv1.Deployment, oldDeployment *appsv1.Deployment) error {
-	// preserve existing resource requirements.
-	existingResources := make(map[string]corev1.ResourceRequirements)
-	for _, container := range oldDeployment.Spec.Template.Spec.Containers {
-		existingResources[container.Name] = container.Resources
-	}
-	// preserve old label selector if it exist, this field is immutable and shouldn't be changed for the lifecycle of the component.
-	if oldDeployment.Spec.Selector != nil {
-		deployment.Spec.Selector = oldDeployment.Spec.Selector.DeepCopy()
-	}
-
-	deploymentConfig, err := c.defaultOptions(cpContext, &deployment.Spec.Template, deployment.Spec.Replicas, existingResources)
+	deploymentConfig, err := c.defaultOptions(cpContext, c.workloadProvider.PodTemplateSpec(workloadObj), c.workloadProvider.Replicas(workloadObj))
 	if err != nil {
 		return err
 	}
-	deploymentConfig.ApplyTo(deployment)
-	return nil
-}
+	c.workloadProvider.ApplyOptionsTo(cpContext, workloadObj, oldWorkloadObj, deploymentConfig)
```

Comment:
I wouldn't expect we need to pass deploymentConfig through any signaturem specially public

Reply:
the problem is there is not generic function for `deploymentConfig.ApplyTo` si each workloadProvider needs to call the correct function. I'd like to get rid of deploymentConfig in the cpov2 completely once finish the refactors to make the code more readable and explicit.

> through any signaturem specially public

technically this should be only used within this package, I can make it the func private

Reply:
we could hide this within the deploymentConfig package rather than surface it in this library
```
	switch any(workloadObj).(type) {
	case appsv1.Deployment:
		// call right function
	case stateful:
```

not a blocker for me

Reply:
thanks, I will follow-up with a PR to remove deploymentConfig from v2